tests/storage/test_base.py

```diff
@@ -209,3 +209,63 @@ def test_delete_one(self) -> Generator["defer.Deferred[object]", object, None]:
         self.mock_txn.execute.assert_called_with(
             "DELETE FROM tablename WHERE keycol = ?", ["Go away"]
         )
+
+    @defer.inlineCallbacks
+    def test_upsert(self) -> Generator["defer.Deferred[object]", object, None]:
```

Comment:
There's some stuff in synapse about "emulated upserts". Do we want to add tests for those?

(Aside: my view would be that we should try to rip out that logic. I think we would need to enforce that background updates which make the necessary unique indices have run?)

Reply:
> There's some stuff in synapse about "emulated upserts". Do we want to add tests for those?

It would probably be good to do this, yeah. Agreed that it would be better to *not* have that though. 😄 See #16397 maybe?